### ISSUE TITLE
Computed value of lineHeight and inheritance.

### ISSUE THREAD

**Issue by palemieux** [w3c/ttml1#220]
Section 8.4.4.2(6)(b) specifies that `lineHeight` is set to its initial value of `"normal"` on `<region>`.

Section 8.4.4.2(6)(c) specifies that `<body>` will inherit the _computed value_ of `lineHeight` from its parent `<region>`.

The computed value of `lineHeight` on `<region>` is therefore, say, 125% of the computed value of `fontSize` on `<region>`, so `125/15rh`.

So the computed values of `lineHeight` and `fontSize` on `<body>` are `125/15rh` and `400/15rh`, respectively.

These computed values are then propagated down by inheritance to the `<p>`, where "Hello" and "Bye" will overlap since the font size is significantly larger than the line height.

Is this the intended outcome and correct interpretation, or alternatively should initial values only be applied to elements to which the styles applies or does "normal" have special inheritance rules?

```
<?xml version="1.0" encoding="UTF-8"?>
<tt xmlns="http://www.w3.org/ns/ttml" xmlns:tts="http://www.w3.org/ns/ttml#styling">
    <head>
        <layout>
            <region xml:id="bottom" tts:origin="10% 10%" tts:extent="80% 80%" />
        </layout>
    </head>
    <body tts:fontSize="4c">
        <div>
            <p region="bottom" begin="00:00:10.000" end="00:00:20.000">Hello<br/>Bye</p>
        </div>
    </body>
</tt>
```

NOTE: `100rh` equals the computed height of root container.

**Comment by skynavga** [comment]
> The computed value of lineHeight on <region> is, say, 125% of the computed value of fontSize on <region>, so 1.25/15rh.

Don't you mean 1.25 \* (100/15)rh, assuming that the font size on region is 1c and that cell resolution is the default of 32 columns by 15 rows. Note that 1rh is not the computed height of root container region; rather, 100rh is the computed height.

**Comment by palemieux** [comment]
> Note that 1rh is not the computed height of root container region; rather, 100rh is the computed height.

Oops. I missed the fact that '1rh` was 1% of the root container height.

If 100rh is the computed height, than "1c" would be "100rh/15" since there are 15 rows, right?

... and 125% of "1c" would be "125rh/15".

P.S.: I updated the bug description

**Comment by skynavga** [comment]
> So the computed values of lineHeight and fontSize on <body> are 125/15rh and 400/15rh, respectively.

I'm not sure where you are getting 400/15rh. Let's start from the beginning:
1. computed cell size (for default cell resolution) is (100/32)rw by (100/15)rh, or [3.125rw,6.667rh];
2. therefore, computed font size on region (if not explicitly specified) is (100/15)rh = 6.667rh;
3. therefore, computed value of line size on region (if not explicitly specified) is 125% \* 6.667rh = 125% \* (100/15)rh = (125/15)rh = 8.334rh.

So, we have:
- computed value of font size on region (inherited by body) of 6.667rh = (100/15)rh, and
- computed value of line height on region (inherited by body) of 8.334rh = (125/15)rh.

So, your suggestion that font size will be larger than line height is incorrect.

**Comment by palemieux** [comment]
> So, your suggestion that font size will be larger than line height is incorrect.

See `tts:fontSize="4c"` on `<body>`

**Comment by skynavga** [comment]
Sorry, missed that (and that explains your 400/15). I see your point now. It appears that we need to require that the computed value of 'normal' remains 'normal', so that normal is only resolved to a _used value_ [1] in the context of its local element's font size.

[1] https://www.w3.org/TR/CSS22/cascade.html#used-value

We also need to add the concept of _used value_ to [2].

[2] http://w3c.github.io/ttml2/spec/ttml2.html#semantics-style-resolution-value-categories

**Comment by palemieux** [comment]
The issue is not limited to "normal". What will be the computed `lineHeight` on `<p>` in the following case, where `lineHeight` is explicitly set to 125% on `<region>`:

```
<?xml version="1.0" encoding="UTF-8"?>
<tt xmlns="http://www.w3.org/ns/ttml" xmlns:tts="http://www.w3.org/ns/ttml#styling">
    <head>
        <layout>
            <region xml:id="bottom" tts:origin="10% 10%" tts:extent="80% 80%" tts:lineHeight="125%" />
        </layout>
    </head>
    <body tts:fontSize="4c">
        <div>
            <p region="bottom" begin="00:00:10.000" end="00:00:20.000">Hello<br/>Bye</p>
        </div>
    </body>
</tt>
```

**Comment by skynavga** [comment]
Yes.

**Comment by palemieux** [comment]
Yes... what is the expected computed value of `lineHeight` on `<p>` in this second example?

**Comment by skynavga** [comment]
Before I answer, I want to check behavior of CSS engines to see whether they do not resolve percentage or normal at the point in the inheritance hierarchy where they are specified, as opposed to resolving on the element which makes use of the property.

**Comment by palemieux** [comment]
Ok. See example at https://developer.mozilla.org/en-US/docs/Web/CSS/line-height, which seems to match the behavior currently specified in TTML.

**Comment by nigelmegitt** [comment]
By the way, the issue very likely applies in cases _other_ than `tts:lineHeight="normal"` due to this text:

> If the value of this attribute is `normal`, then the computed value of the style property must be considered to be no less than the size of the largest font size that applies to the element and its descendant elements in the intermediate synchronic document as determined by 9.3.2 Intermediate Synchronic Document Construction.

That says to me that in the original example in the issue the computed value must be no less than 4c for the `p` element.

For % values I agree that this may give a _surprising_ result but I do not think it is necessarily an incorrect result. Deferring evaluation of computed % values to the `p` element would be equally surprising in the context of inheritance generally applying to computed values, and also there's an easy workaround, which is to specify `tts:lineHeight` directly on the `p` if that is what is required.

By the way, the text above about `normal` could give strange behaviours, in that the line spacing could change between ISDs if the only difference between ISDs is the inclusion/exclusion of text with a larger font size than all the other text. This would look like baselines jumping around, in the case that there is other text that is common across such an ISD boundary. It is also computationally more complex to depend on ISDs from the perspective of a transformation processor. It would be easier to depend on all descendant elements regardless of ISD; that would require computed style evaluation of `tts:lineHeight` before ISD creation.

For example, adjusting Pierre's initial example:

``` xml
<?xml version="1.0" encoding="UTF-8"?>
<tt xmlns="http://www.w3.org/ns/ttml" xmlns:tts="http://www.w3.org/ns/ttml#styling">
    <head>
        <layout>
            <region xml:id="bottom" tts:origin="10% 10%" tts:extent="80% 80%"/>
        </layout>
    </head>
    <body tts:fontSize="4c" tts:lineHeight="normal">
        <div region="bottom">
            <p begin="00:00:10.000" end="00:00:20.000">Hello<br/>Bye</p>
            <p begin="00:00:15.000" end="00:00:20.000" tts:fontSize="200%">Double size text</p>
        </div>
    </body>
</tt>
```

In this example, at 15s the spacing between "Hello" and "Bye" suddenly doubles, which is certainly counter-intuitive.

**Comment by nigelmegitt** [comment]
One more thought on the `lineHeight="normal"` text above: it does not qualify the descendants to exclude any that specify `lineHeight`, which means that regardless of any `lineHeight` settings on descendants their `fontSize` _is_ taken into account when computing the value `normal`. It may be nice to be able to exclude parts of the descendant tree based on some criteria when specifying `lineHeight`.

**Comment by palemieux** [comment]
> If the value of this attribute is normal, then the computed value of the style property must be considered to be no less than the size of the largest font size that
> applies to the element and its descendant elements in the intermediate synchronic document as determined by 9.3.2 Intermediate Synchronic Document Construction.

I agree this text is problematic, and am starting to think that TTML could simply recommend that "normal" = "125%", recommend that lineHeight always be specified whenever fontSize is specified, and let the style resolution rules apply as currently specified:
- it yields predictable results and is easy to compute
- it is closer to the current TTML2 draft
- matches CSS AFAIK

Other approaches seem to involve modifying the style resolution process and/or introducing a new metric like CSS does.

**Comment by nigelmegitt** [comment]
I'm afraid I wouldn't be happy with that proposal @palemieux . It's important to have a semantic that allows for automated sensible baseline positioning of multiple font size content, and a single percentage value doesn't do it, unless you intend to add some wording about the basis of the 125% being the largest descendant content's computed font size, perhaps with some exceptions for content elements that specify a different lineHeight. That is what the current wording is trying to achieve I believe, though not ideally. 

I do not have a counter proposal right now - whatever we do should probably align with CSS though, as a starting point. When you try to render using a mapping to HTML/CSS you'll find that no single percentage value works across all browsers, so some kind of flexibility is possibly worth retaining.

Notes: 

CSS reference: https://www.w3.org/TR/CSS2/visudet.html#propdef-line-height

This includes:

> normal
> Tells user agents to set the used value to a "reasonable" value based on the font of the element. The value has the same meaning as <number>. We recommend a used value for 'normal' between 1.0 to 1.2. The computed value is 'normal'.

and

> When an element contains text that is rendered in more than one font, user agents may determine the 'normal' 'line-height' value according to the largest font size.

Interestingly the computed value is actually the keyword in CSS, which avoids the computed value inheritance issue identified in this issue, effectively delegating computation of the numerical value until it is needed on an element. It is also interesting to me that CSS is permissive of using the largest font size but does not require it; that does not seem like a good fit for TTML.

**Comment by palemieux** [comment]
>  It's important to have a semantic that allows for automated sensible baseline positioning of multiple font size content, and a single percentage value doesn't do it, 

@nigelmegitt In what circumstances would the author not be able to set a lineHeight explicitly?

**Comment by nigelmegitt** [comment]
> In what circumstances would the author not be able to set a lineHeight explicitly?

Logically of course it is possible to set a value; however there is no way to guarantee that this is a _good_ value that results in readable text with a sensible spacing. This is what the updated algorithm in TTML2 helps with, and what the CSS "normal" value permits; since you cannot in general know which actual font is going to be used and what the appropriate line height will be for it.

In some use cases line positioning needs to be tightly controlled; in others there is more freedom and line spacing can be varied dependent on font choice. We should allow for both.

**Comment by palemieux** [comment]
> however there is no way to guarantee that this is a good value that results in readable text with a sensible spacing.

Can you provide examples of where setting lineHeight to 125% of the font size does not yield acceptable results?

> This is what the updated algorithm in TTML2 helps with,

The algorithm in the TTML2 ED only takes into account the fontSize on the element itself and not on its descendants.

**Comment by nigelmegitt** [comment]
> Can you provide examples of where setting lineHeight to 125% of the font size does not yield acceptable results?

Where it generates a spacing that is actually too large for the font and looks strange. Note that the CSS recommendation is a value between 1.0 and 1.2.

> The algorithm in the TTML2 ED only takes into account the fontSize on the element itself and not on its descendants.

Yes, I just meant that it is helpful for one part of the problem - I agree it should take into account descendants somehow too.

**Comment by palemieux** [comment]
> Where it generates a spacing that is actually too large for the font and looks strange.

Is this an hypothetical scenario, or do you have a specific example?

**Comment by nigelmegitt** [comment]
I think this is pretty common actually. Few combinations of font and font-size generate a normal spacing of 125% in most browsers, but the text generally reads fine. So making the line height bigger would probably look a bit odd, or a bit "different" anyway.

**Comment by nigelmegitt** [comment]
Here's another piece of the puzzle: 

https://developer.mozilla.org/en/docs/Web/CSS/line-height points out that avoiding inheritance problems is often achieved by using a dimensionless number which acts as a multiplier locally on the element's fontSize rather than changing the computed value directly. This can be seen in action at https://jsfiddle.net/rs02y18z/

In TTML we do not have the concept of a dimensionless number that can be passed down and applied locally at all - we seem only to have inheritance of computed values. There may be an argument for adding this to TTML2 since similar issues to those found in HTML/CSS are clearly arising when it comes to lineHeight.

**Comment by skynavga** [comment]
I've checked two implementations of TTML rendering engines (TTPE and DFXPVW), and both use 'normal' as the computed value when 'normal' is the specified value, and, consequently, descendants inherit 'normal', where it is finally evaluated in the local element context against font sizes.

I also note that CSS2.1 says that the `normal` value "has the same meaning as `<number>`". Further, `<number>` is defined such that "the computed value is the same as the specified value.".

Now, even though we don't have the option of using `<number>` in TTML1, we still need to interpret `normal` in a manner that is consistent with XSL-FO and CSS, which is to say, the computed value of normal is treated as a numeric multiplier (as if `<number>` were supported).

Note also, that In TTML2, we had previously added an algorithm for computing line height, where the first step is "If the value of this attribute is `normal`, then ... (1) let _P_ be the `p` element to which this style property applies ...". It is clear from this text that there is an expectation that line height might be `normal` on an element like `p`, and further, there is not an expectation or implication that this would require having an explicit `tts:lineHeight="normal"` on `p`. In other words, there is an assumption that `normal` is inherited as itself, i.e., without resolving it to a `<length>`.

To resolve this, I think we need to add material  under tts:lineHeight that describes this special behavior.

**Comment by palemieux** [comment]
> To resolve this, I think we need to add material under tts:lineHeight that describes this special behavior.

To summarize:
- "normal" is inherited as "normal" and computed (separately on each element to which it applies) as a multiple of the computed fontSize on the element
- the computed values of "percentage" and "em" are inherited (as currently specified)

**Comment by nigelmegitt** [comment]
The way I would describe this, the value `normal` is itself inheritable as a computed value, which is only evaluated into a real size when stacking the line areas associated with the `p` element to which it applies.

In terms of actions, at the very least, we need an erratum to TTML1SE explaining that the computed value associated with the specified value of "normal" is "normal".

Re-reading the text:

> If the value of this attribute is normal, then the computed value of the style property must be considered to be no less than the size of the largest font size that applies to the element and its descendant elements in the intermediate synchronic document as determined by 9.3.2 Intermediate Synchronic Document Construction.

I see that this does not actually preclude using the largest font size of all descendants; it may be worth adding by errata an informative note to explain this and that it can be used to avoid baselines jumping at ISD junctions, which can still occur if a span is introduced with larger-sized text, since "normal" is presumably only evaluated on the `p` element to which `tts:lineHeight` applies. For TTML2 I would prefer to see an even more clearly specified rule however, while still retaining the ability to size based on the largest font size of the `p`'s contents.

In summary, I think there are two actions for TTML1, both in an erratum:
- [ ] Clarify that the computed value of `normal` is `normal` up until it has to be evaluated to an actual size, on the `p` element.
- [ ] Informatively note that the largest font size of all descendants of the `p` can be used, to avoid baseline jumping.

Also for TTML2:
- [ ] Capture these changes and any further implications for TTML2 as issues.

**Comment by palemieux** [comment]
> I see that this does not actually preclude using the largest font size of all descendants; it may be worth adding by errata an informative note
> to explain this and that it can be used to avoid baselines jumping at ISD junctions, 

How would baseline jumping be avoided since styles are resolved and computed independently for each ISD?

**Comment by nigelmegitt** [comment]
> How would baseline jumping be avoided since styles are resolved and computed independently for each ISD?

Because the rule is _no less than the size of the largest font size..._ - this means that computing the lineHeight for **all** descendants regardless of ISD also satisfies the rule. It does not additionally require "shall be no larger than the size of the largest font size..." which would carry that implication.

**Comment by palemieux** [comment]
> Because the rule is no less than the size of the largest font size... - this means that computing the lineHeight for all descendants regardless of ISD also satisfies the rule. 

IMHO this is a significant departure from the ISD and style resolution algorithms specified in TTML, as well as additional complexity for the player, to address a situation that can be easily remedied by the author specifying an explicit line height on `<p>`, which is presumably always possible.

**Comment by nigelmegitt** [comment]
> IMHO this is a significant departure from the ISD and style resolution algorithms specified in TTML, as well as additional complexity for the player, to address a situation that can be easily remedied by the author specifying an explicit line height on <p>, which is presumably always possible. 

Well it is not a requirement, just a permitted implementation option. For TTML1 I don't see a problem with this, but as I said, for TTML2 I would prefer to tighten this up. By the way, I think we are discussing a rarely encountered detail.

**Comment by palemieux** [comment]
Ok. Wouldn't it make sense to add to your checklist above: "recommend that authors set lineHeight explicitly on `<p>` for consistent presentation"?

**Comment by nigelmegitt** [comment]
> Wouldn't it make sense to add to your checklist above: "recommend that authors set lineHeight explicitly on `<p>` for consistent presentation"?

If you mean setting an explicit value as opposed to `normal`, I wouldn't agree with that as a recommendation since only `normal` gives implementations the freedom to inspect the actually used font's metrics to generate a nice-looking line spacing. I agree there are trade-offs both ways but making that a recommendation seems too strong to me. I would not mind an informative note pointing out the pros and cons of the different approaches.

**Comment by nigelmegitt** [comment]
[Meeting 2016-11-17] @skynavga unsure if the baseline jumping issue is a correct interpretation of the text, will review.

**Comment by nigelmegitt** [comment]
[Meeting 2017-01-12] Agreed to specify that the value "normal" is inherited prior to computing line height, rather than the computed value being inherited.
